tests/test_appipe.py

```diff
@@ -21,22 +21,142 @@
+                            yield pipeBase.Struct(ccdProcessor=mockCcdProcessor,
+                                                  differencer=mockDifferencer,
+                                                  dpddifier=mockDpddifier,
+                                                  associator=mockAssociator,
+                                                  forcedSource=mockForcedSource)
+
+    def testGenericRun(self):
+        # test the normal workflow of each ap_pipe step
+        task = ApPipeTask(self.butler, config=self.config)
+        inputRef = self.butler.dataRef("raw", **self.dataId)
+        with self.mockPatchSubtasks(task) as subtasks:
+            task.runDataRef(inputRef)
+            subtasks.ccdProcessor.runDataRef.assert_called_once()
+            subtasks.differencer.runDataRef.assert_called_once()
+            subtasks.associator.run.assert_called_once()
+            subtasks.forcedSource.run.assert_called_once()
+
+    def testReuseExistingOutput(self):
+        # test reuse keyword to ApPipeTask.runDataRef
+        task = ApPipeTask(self.butler, config=self.config)
+        inputRef = self.butler.dataRef("raw", **self.dataId)
+
+        # Mock guarantees that all "has this been made" tests pass,
+        # so skippable subtasks should actually be skipped
+        for skippable in [
+            ['ccdProcessor'],
+            ['ccdProcessor', 'differencer'],
+        ]:
+            # Start with a fresh mock for each test case
+            with self.mockPatchSubtasks(task) as subtasks:
+                struct = task.runDataRef(inputRef, reuse=skippable)
+                for subtaskName, runner in {
+                    'ccdProcessor': subtasks.ccdProcessor.runDataRef,
+                    'differencer': subtasks.differencer.runDataRef,
+                    'associator': subtasks.associator.run,
+                }.items():
+                    if subtaskName in skippable:
+                        runner.assert_not_called()
+                        self.assertIsNone(struct.getDict()[subtaskName])
+                    else:
+                        runner.assert_called_once()
+                        self.assertIsNotNone(struct.getDict()[subtaskName])
+
+    def testCalexpRun(self):
+        # test the calexp template workflow of each ap_pipe step
+        calexpConfigFile = os.path.join(lsst.utils.getPackageDir('ap_pipe'),
+                                        'config', 'calexpTemplates.py')
+        calexpConfig = self._makeDefaultConfig()
+        calexpConfig.load(calexpConfigFile)
+        calexpConfig.differencer.doSelectSources = False  # Workaround for DM-18394
 
-#    def testDoDiffIm(self):
-#        # test something
+        task = ApPipeTask(self.butler, config=calexpConfig)
+        inputRef = self.butler.dataRef("raw", **self.dataId)
+        with self.mockPatchSubtasks(task) as subtasks:
+            # Self-subtraction ok because we're not actually doing any processing
```

Comment:
I don't understand what this comment is referring to?

Reply:
Clarified. It was referring to the fact that the differencing template has the same dataId as the science template.